File: watchos/Models/Expression.swift
// Expression parsing and evaluation — the Swift replacement for mathjs.
//
// Direct copy of ios/App/Expression.swift. The watch app has no equation input (no
// practical keyboard for `sin(x)` on a 40mm screen), so it only ever compiles the fixed
// preset list in Presets.swift, but it needs the real evaluator to plot them rather than
// faking curve shapes. Keep this file byte-for-byte in sync with the iOS original.
//
// The web app calls mathjs `parse(expr).compile()` and evaluates it per pixel with one
// variable, x (see src/utils/evaluate.js). That is all it ever asks of mathjs, so this is a
// recursive-descent parser for exactly that grammar rather than a general CAS.
//
// Grammar (lowest to highest precedence):
//   expression := term (('+' | '-') term)*
//   term       := unary (('*' | '/') unary | implicit-multiplication)*
//   unary      := ('-' | '+')* power
//   power      := primary ('^' unary)?          -- right associative
//   primary    := number | 'x' | constant | function '(' expression ')' | '(' expression ')'

import Foundation

enum ExpressionError: Error {
    case message(String)

    var text: String {
        switch self {
        case .message(let m): return m
        }
    }
}

// MARK: - Tokens

private enum Token: Equatable {
    case number(Double)
    case identifier(String)
    case op(Character)
    case lparen
    case rparen
}

private func tokenize(_ input: String) throws -> [Token] {
    var tokens: [Token] = []
    let chars = Array(input)
    var i = 0

    while i < chars.count {
        let c = chars[i]

        if c.isWhitespace { i += 1; continue }

        if c.isNumber || c == "." {
            var literal = ""
            while i < chars.count, chars[i].isNumber || chars[i] == "." {
                literal.append(chars[i]); i += 1
            }
            guard let value = Double(literal) else {
                throw ExpressionError.message("Not a number: \(literal)")
            }
            tokens.append(.number(value))
            continue
        }

        if c.isLetter {
            var name = ""
            while i < chars.count, chars[i].isLetter || chars[i].isNumber {
                name.append(chars[i]); i += 1
            }
            tokens.append(.identifier(name.lowercased()))
            continue
        }

        switch c {
        case "+", "-", "*", "/", "^":
            tokens.append(.op(c)); i += 1
        case "(":
            tokens.append(.lparen); i += 1
        case ")":
            tokens.append(.rparen); i += 1
        default:
            throw ExpressionError.message("Unexpected character: \(c)")
        }
    }
    return tokens
}

// MARK: - Syntax tree

private indirect enum Node {
    case constant(Double)
    case variable
    case unaryMinus(Node)
    case binary(Character, Node, Node)
    case call(String, Node)
}

private nonisolated(unsafe) let functions: [String: (Double) -> Double] = [
    "sin": sin, "cos": cos, "tan": tan,
    "asin": asin, "acos": acos, "atan": atan,
    "sinh": sinh, "cosh": cosh, "tanh": tanh,
    "sqrt": sqrt, "cbrt": cbrt, "abs": abs,
    // mathjs `log` is the natural logarithm — keep that, or every log() plot shifts.
    "log": Foundation.log, "ln": Foundation.log,
    "log10": log10, "log2": log2, "exp": exp,
    "floor": floor, "ceil": ceil, "round": { $0.rounded() },
    "sign": { $0 > 0 ? 1 : ($0 < 0 ? -1 : 0) },
]

private let constants: [String: Double] = [
    "pi": .pi, "e": M_E, "tau": 2 * .pi,
]

private struct Parser {
    let tokens: [Token]
    var position = 0

    var current: Token? { position < tokens.count ? tokens[position] : nil }

    mutating func parse() throws -> Node {
        let node = try expression()
        if position < tokens.count {
            throw ExpressionError.message("Unexpected trailing input")
        }
        return node
    }

    mutating func expression() throws -> Node {
        var left = try term()
        while case .op(let c) = current, c == "+" || c == "-" {
            position += 1
            left = .binary(c, left, try term())
        }
        return left
    }

    mutating func term() throws -> Node {
        var left = try unary()
        while true {
            if case .op(let c) = current, c == "*" || c == "/" {
                position += 1
                left = .binary(c, left, try unary())
                continue
            }
            // Implicit multiplication: mathjs accepts `2x` and `2sin(x)`, and people type them.
            if startsPrimary(current) {
                left = .binary("*", left, try unary())
                continue
            }
            return left
        }
    }

    private func startsPrimary(_ token: Token?) -> Bool {
        switch token {
        case .number, .identifier, .lparen: return true
        default: return false
        }
    }

    mutating func unary() throws -> Node {
        if case .op(let c) = current, c == "-" || c == "+" {
            position += 1
            let operand = try unary()
            return c == "-" ? .unaryMinus(operand) : operand
        }
        return try power()
    }

    mutating func power() throws -> Node {
        let base = try primary()
        if case .op(let c) = current, c == "^" {
            position += 1
            return .binary("^", base, try unary()) // right associative: 2^3^2 == 2^9
        }
        return base
    }

    mutating func primary() throws -> Node {
        guard let token = current else {
            throw ExpressionError.message("Unexpected end of expression")
        }

        switch token {
        case .number(let value):
            position += 1
            return .constant(value)

        case .lparen:
            position += 1
            let inner = try expression()
            guard case .rparen = current else {
                throw ExpressionError.message("Missing closing parenthesis")
            }
            position += 1
            return inner

        case .identifier(let name):
            position += 1
            if name == "x" { return .variable }
            if let value = constants[name] { return .constant(value) }
            guard functions[name] != nil else {
                throw ExpressionError.message("Unknown name: \(name)")
            }
            guard case .lparen = current else {
                throw ExpressionError.message("\(name) needs parentheses, like \(name)(x)")
            }
            position += 1
            let argument = try expression()
            guard case .rparen = current else {
                throw ExpressionError.message("Missing closing parenthesis after \(name)(")
            }
            position += 1
            return .call(name, argument)

        case .rparen:
            throw ExpressionError.message("Unmatched closing parenthesis")

        case .op(let c):
            throw ExpressionError.message("Unexpected operator: \(c)")
        }
    }
}

private func evaluate(_ node: Node, x: Double) -> Double {
    switch node {
    case .constant(let value):
        return value
    case .variable:
        return x
    case .unaryMinus(let operand):
        return -evaluate(operand, x: x)
    case .call(let name, let argument):
        return functions[name]?(evaluate(argument, x: x)) ?? .nan
    case .binary(let op, let lhs, let rhs):
        let a = evaluate(lhs, x: x)
        let b = evaluate(rhs, x: x)
        switch op {
        case "+": return a + b
        case "-": return a - b
        case "*": return a * b
        case "/": return a / b
        case "^": return pow(a, b)
        default: return .nan
        }
    }
}

// MARK: - Public surface, mirroring src/utils/evaluate.js

struct CompiledExpression {
    /// nil when the input was empty — an empty row is not an error, it just draws nothing.
    let function: ((Double) -> Double)?
    let error: String?
}

func compileExpression(_ raw: String) -> CompiledExpression {
    // Strip a leading `y =`, same as the web version's /^y\s*=\s*/i.
    var cleaned = raw.trimmingCharacters(in: .whitespaces)
    if let match = cleaned.range(of: #"^y\s*=\s*"#, options: [.regularExpression, .caseInsensitive]) {
        cleaned = String(cleaned[match.upperBound...])
    }
    cleaned = cleaned.trimmingCharacters(in: .whitespaces)

    guard !cleaned.isEmpty else { return CompiledExpression(function: nil, error: nil) }

    do {
        let tokens = try tokenize(cleaned)
        var parser = Parser(tokens: tokens)
        let tree = try parser.parse()
        return CompiledExpression(function: { evaluate(tree, x: $0) }, error: nil)
    } catch let error as ExpressionError {
        return CompiledExpression(function: nil, error: error.text)
    } catch {
        return CompiledExpression(function: nil, error: "Could not read that expression")
    }
}

/// A vertical asymptote (tan(x), 1/x) yields two finite but huge values of opposite sign on
/// adjacent pixels. Joining them draws a false vertical line, so the stroke has to break.
func isAsymptoteJump(previous: Double, current: Double, scale: Double, height: Double) -> Bool {
    // Mirrors Math.sign() exactly, zero included, so the Swift and JS versions break in the
    // same places.
    func sign(_ value: Double) -> Double { value > 0 ? 1 : (value < 0 ? -1 : 0) }
    return sign(current) != sign(previous) && abs(current - previous) * scale > height * 2
}

File: kmp/shared/src/commonMain/kotlin/com/nulljosh/curvely/Expression.kt
package com.nulljosh.curvely

import kotlin.math.*

// Ported from ios/App/Expression.swift (itself the Swift replacement for
// mathjs). Keep the three in lockstep: src/utils/evaluate.js (web),
// ios/App/Expression.swift (Swift), this file (Kotlin).

class ExpressionException(message: String) : Exception(message)

private sealed class Tok {
    data class Number(val value: Double) : Tok()
    data class Ident(val name: String) : Tok()
    data class Op(val c: Char) : Tok()
    object LParen : Tok(); object RParen : Tok()
}

private fun tokenize(input: String): List<Tok> {
    val tokens = mutableListOf<Tok>()
    val chars = input.toCharArray()
    var i = 0
    while (i < chars.size) {
        val c = chars[i]
        if (c.isWhitespace()) { i++; continue }
        if (c.isDigit() || c == '.') {
            val start = i
            while (i < chars.size && (chars[i].isDigit() || chars[i] == '.')) i++
            val literal = String(chars, start, i - start)
            val value = literal.toDoubleOrNull() ?: throw ExpressionException("Not a number: $literal")
            tokens.add(Tok.Number(value)); continue
        }
        if (c.isLetter()) {
            val start = i
            while (i < chars.size && (chars[i].isLetterOrDigit())) i++
            tokens.add(Tok.Ident(String(chars, start, i - start).lowercase())); continue
        }
        when (c) {
            '+', '-', '*', '/', '^' -> tokens.add(Tok.Op(c))
            '(' -> tokens.add(Tok.LParen)
            ')' -> tokens.add(Tok.RParen)
            else -> throw ExpressionException("Unexpected character: $c")
        }
        i++
    }
    return tokens
}

private sealed class Node {
    data class Const(val value: Double) : Node()
    object Variable : Node()
    data class UnaryMinus(val operand: Node) : Node()
    data class Binary(val op: Char, val lhs: Node, val rhs: Node) : Node()
    data class Call(val name: String, val arg: Node) : Node()
}

private val FUNCTIONS: Map<String, (Double) -> Double> = mapOf(
    "sin" to ::sin, "cos" to ::cos, "tan" to ::tan,
    "asin" to ::asin, "acos" to ::acos, "atan" to ::atan,
    "sinh" to ::sinh, "cosh" to ::cosh, "tanh" to ::tanh,
    "sqrt" to ::sqrt, "cbrt" to ::cbrt, "abs" to ::abs,
    "log" to ::ln, "ln" to ::ln,
    "log10" to ::log10, "log2" to { x: Double -> log2(x) }, "exp" to ::exp,
    "floor" to ::floor, "ceil" to ::ceil, "round" to { x: Double -> round(x) },
    "sign" to { x: Double -> sign(x) },
)

private val CONSTANTS: Map<String, Double> = mapOf("pi" to PI, "e" to E, "tau" to 2 * PI)

private class Parser(private val tokens: List<Tok>) {
    var position = 0
    private fun current(): Tok? = tokens.getOrNull(position)

    fun parse(): Node {
        val node = expression()
        if (position < tokens.size) throw ExpressionException("Unexpected trailing input")
        return node
    }

    private fun startsPrimary(t: Tok?) = t is Tok.Number || t is Tok.Ident || t is Tok.LParen

    fun expression(): Node {
        var left = term()
        while (true) {
            val t = current()
            if (t is Tok.Op && (t.c == '+' || t.c == '-')) { position++; left = Node.Binary(t.c, left, term()) }
            else return left
        }
    }

    private fun term(): Node {
        var left = unary()
        while (true) {
            val t = current()
            if (t is Tok.Op && (t.c == '*' || t.c == '/')) { position++; left = Node.Binary(t.c, left, unary()); continue }
            if (startsPrimary(t)) { left = Node.Binary('*', left, unary()); continue }
            return left
        }
    }

    private fun unary(): Node {
        val t = current()
        if (t is Tok.Op && (t.c == '-' || t.c == '+')) {
            position++
            val operand = unary()
            return if (t.c == '-') Node.UnaryMinus(operand) else operand
        }
        return power()
    }

    private fun power(): Node {
        val base = primary()
        val t = current()
        if (t is Tok.Op && t.c == '^') { position++; return Node.Binary('^', base, unary()) }
        return base
    }

    private fun primary(): Node {
        val t = current() ?: throw ExpressionException("Unexpected end of expression")
        return when (t) {
            is Tok.Number -> { position++; Node.Const(t.value) }
            is Tok.LParen -> {
                position++
                val inner = expression()
                if (current() !is Tok.RParen) throw ExpressionException("Missing closing parenthesis")
                position++
                inner
            }
            is Tok.Ident -> {
                position++
                when {
                    t.name == "x" -> Node.Variable
                    t.name in CONSTANTS -> Node.Const(CONSTANTS.getValue(t.name))
                    t.name in FUNCTIONS -> {
                        if (current() !is Tok.LParen) throw ExpressionException("${t.name} needs parentheses, like ${t.name}(x)")
                        position++
                        val arg = expression()
                        if (current() !is Tok.RParen) throw ExpressionException("Missing closing parenthesis after ${t.name}(")
                        position++
                        Node.Call(t.name, arg)
                    }
                    else -> throw ExpressionException("Unknown name: ${t.name}")
                }
            }
            is Tok.RParen -> throw ExpressionException("Unmatched closing parenthesis")
            is Tok.Op -> throw ExpressionException("Unexpected operator: ${t.c}")
        }
    }
}

private fun evaluate(node: Node, x: Double): Double = when (node) {
    is Node.Const -> node.value
    is Node.Variable -> x
    is Node.UnaryMinus -> -evaluate(node.operand, x)
    is Node.Call -> FUNCTIONS[node.name]?.invoke(evaluate(node.arg, x)) ?: Double.NaN
    is Node.Binary -> {
        val a = evaluate(node.lhs, x)
        val b = evaluate(node.rhs, x)
        when (node.op) {
            '+' -> a + b; '-' -> a - b; '*' -> a * b; '/' -> a / b; '^' -> a.pow(b)
            else -> Double.NaN
        }
    }
}

data class CompiledExpression(val function: ((Double) -> Double)?, val error: String?)

private val Y_PREFIX = Regex("^y\\s*=\\s*", RegexOption.IGNORE_CASE)

fun compileExpression(raw: String): CompiledExpression {
    var cleaned = raw.trim().replace(Y_PREFIX, "").trim()
    if (cleaned.isEmpty()) return CompiledExpression(null, null)
    return try {
        val tokens = tokenize(cleaned)
        val tree = Parser(tokens).parse()
        CompiledExpression({ x -> evaluate(tree, x) }, null)
    } catch (e: ExpressionException) {
        CompiledExpression(null, e.message)
    } catch (e: Exception) {
        CompiledExpression(null, "Could not read that expression")
    }
}

/** A vertical asymptote (tan(x), 1/x) yields two finite but huge values of opposite sign on
 *  adjacent pixels. Joining them draws a false vertical line, so the stroke has to break. */
fun isAsymptoteJump(previous: Double, current: Double, scale: Double, height: Double): Boolean {
    fun s(v: Double) = if (v > 0) 1.0 else if (v < 0) -1.0 else 0.0
    return s(current) != s(previous) && abs(current - previous) * scale > height * 2
}
